core/ctree/cnode.cpp

```diff
@@ -0,0 +1,420 @@
+        int is_reset = 0;
+        while(node_stack.size() > 0){
+            CNode* node = node_stack.top();
+            node_stack.pop();
+
+            if(node != root){
+                float true_reward = node->value_prefix - parent_value_prefix;
+                if(is_reset == 1){
+                    true_reward = node->value_prefix;
+                }
+                float qsa = true_reward + discount * node->value();
+                min_max_stats.update(qsa);
+            }
+
+            for(int a = 0; a < node->action_num; ++a){
+                CNode* child = node->get_child(a);
+                if(child->expanded()){
+                    node_stack.push(child);
+                }
+            }
+
+            parent_value_prefix = node->value_prefix;
+            is_reset = node->is_reset;
+        }
+    }
+
+    void cback_propagate(std::vector<CNode*> &search_path, tools::CMinMaxStats &min_max_stats, int to_play, float value, float discount){
+        float bootstrap_value = value;
+        int path_len = search_path.size();
+        for(int i = path_len - 1; i >= 0; --i){
+            CNode* node = search_path[i];
+            node->value_sum += bootstrap_value;
+            node->visit_count += 1;
+
+            float parent_value_prefix = 0.0;
+            int is_reset = 0;
+            if(i >= 1){
+                CNode* parent = search_path[i - 1];
+                parent_value_prefix = parent->value_prefix;
+                is_reset = parent->is_reset;
+//                float qsa = (node->value_prefix - parent_value_prefix) + discount * node->value();
+//                min_max_stats.update(qsa);
+            }
+
+            float true_reward = node->value_prefix - parent_value_prefix;
+            if(is_reset == 1){
+                // parent is reset
+                true_reward = node->value_prefix;
+            }
+
+            bootstrap_value = true_reward + discount * bootstrap_value;
+        }
+        min_max_stats.clear();
+        CNode* root = search_path[0];
+        update_tree_q(root, min_max_stats, discount);
+    }
+
+    void cbatch_back_propagate(int hidden_state_index_x, float discount, const std::vector<float> &value_prefixs, const std::vector<float> &values, const std::vector<std::vector<float>> &policies, tools::CMinMaxStatsList *min_max_stats_lst, CSearchResults &results, std::vector<int> is_reset_lst){
+        for(int i = 0; i < results.num; ++i){
+            results.nodes[i]->expand(0, hidden_state_index_x, i, value_prefixs[i], policies[i]);
+            // reset
+            results.nodes[i]->is_reset = is_reset_lst[i];
+
+            cback_propagate(results.search_paths[i], min_max_stats_lst->stats_lst[i], 0, values[i], discount);
+        }
+    }
+
+    int cselect_child(CNode* root, tools::CMinMaxStats &min_max_stats, int pb_c_base, float pb_c_init, float discount, float mean_q){
+        float max_score = FLOAT_MIN;
+        const float epsilon = 0.000001;
```

Comment:
This epislon I think is different in our code.

Reply:
I dont actually think there is an epsilon in our code anywhere